File: frontend/src/features/workflow-admin/models/ApprovalMatrix.ts
/**
 * Approval matrix types.
 * Operator and data-type unions mirror the backend enums, so an invalid rule
 * cannot be constructed in the UI in the first place.
 */

export const RULE_OPERATORS = [
  'EQ',
  'NEQ',
  'GT',
  'GTE',
  'LT',
  'LTE',
  'IN',
  'NOT_IN',
  'CONTAINS',
  'STARTS_WITH',
] as const;

export type RuleOperator = (typeof RULE_OPERATORS)[number];

export const RULE_DATA_TYPES = ['STRING', 'NUMBER', 'BOOLEAN', 'LIST'] as const;

export type RuleDataType = (typeof RULE_DATA_TYPES)[number];

export const ASSIGNMENT_TYPES = ['ROLE', 'USER'] as const;

export type AssignmentType = (typeof ASSIGNMENT_TYPES)[number];

export type ApprovalTaskStatus = 'PENDING' | 'COMPLETED' | 'CANCELLED' | 'ESCALATED';

export interface ApprovalRule {
  id: string;
  field: string;
  operator: RuleOperator;
  value: string;
  data_type: RuleDataType;
  logical_group: string;
}

export interface ApprovalAssignment {
  id: string;
  level: number;
  assignment_type: AssignmentType;
  user_id: string | null;
  role_id: string | null;
}

export interface ApprovalMatrix {
  id: string;
  code: string;
  name: string;
  entity_type: string;
  priority: number;
  is_active: boolean;
  rules: ApprovalRule[];
  assignments: ApprovalAssignment[];
  created_by: string;
  created_date: string;
  modified_by: string;
  modified_date: string;
}

export interface ApprovalMatrixListResponse {
  matrices: ApprovalMatrix[];
  total: number;
  skip: number;
  limit: number;
}

/** Rule as submitted — no id, because the backend replaces the whole rule set. */
export interface ApprovalRuleInput {
  field: string;
  operator: RuleOperator;
  value: string;
  data_type: RuleDataType;
  logical_group: string;
}

export interface ApprovalAssignmentInput {
  level: number;
  assignment_type: AssignmentType;
  user_id: string | null;
  role_id: string | null;
}

export interface CreateApprovalMatrixRequest {
  code: string;
  name: string;
  entity_type: string;
  priority: number;
  is_active: boolean;
  rules: ApprovalRuleInput[];
  assignments: ApprovalAssignmentInput[];
}

export type UpdateApprovalMatrixRequest = Partial<
  Omit<CreateApprovalMatrixRequest, 'code'>
>;

export interface ApprovalResolveRequest {
  entity_type: string;
  entity_data: Record<string, unknown>;
}

export interface ApprovalResolveResponse {
  matched: boolean;
  matrix: ApprovalMatrix | null;
  levels: number[];
  assignments: ApprovalAssignment[];
}

export interface ApprovalTask {
  id: string;
  instance_id: string;
  matrix_id: string | null;
  assignee_id: string;
  level: number;
  status: ApprovalTaskStatus;
  action_taken: string | null;
  due_date: string | null;
  comments: string | null;
  created_date: string;
}

File: backend/src/domain/entities/approval_matrix.py
"""Approval matrix domain entities."""

import dataclasses
from dataclasses import dataclass
from datetime import datetime

from src.domain.entities.base_entity import BaseEntity
from src.domain.enums.workflow_enums import (
    ApprovalTaskStatus,
    AssignmentType,
    RuleDataType,
    RuleOperator,
)


@dataclass(kw_only=True)
class ApprovalRule(BaseEntity):
    matrix_id: int | None = None
    field: str = ""
    operator: RuleOperator = RuleOperator.EQ
    value: str = ""
    data_type: RuleDataType = RuleDataType.STRING
    logical_group: str = "default"


@dataclass(kw_only=True)
class ApprovalAssignment(BaseEntity):
    matrix_id: int | None = None
    assignment_type: AssignmentType = AssignmentType.ROLE
    user_id: int | None = None
    role_id: int | None = None
    level: int = 1


@dataclass(kw_only=True)
class ApprovalMatrix(BaseEntity):
    code: str = ""
    name: str = ""
    entity_type: str = ""
    priority: int = 0
    is_active: bool = True
    rules: list[ApprovalRule] = dataclasses.field(default_factory=list)
    assignments: list[ApprovalAssignment] = dataclasses.field(default_factory=list)

    @property
    def levels(self) -> list[int]:
        return sorted({a.level for a in self.assignments})

    def assignments_for_level(self, level: int) -> list[ApprovalAssignment]:
        return [a for a in self.assignments if a.level == level]


@dataclass(kw_only=True)
class ApprovalTask(BaseEntity):
    instance_id: int = 0
    assignee_id: int = 0
    matrix_id: int | None = None
    level: int = 1
    status: ApprovalTaskStatus = ApprovalTaskStatus.PENDING
    action_taken: str | None = None
    due_date: datetime | None = None
    comments: str | None = None

    @property
    def is_open(self) -> bool:
        return self.status == ApprovalTaskStatus.PENDING
